quickwit/quickwit-indexing/src/source/queue_sources/sqs_queue.rs

```diff
@@ -0,0 +1,471 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use anyhow::{bail, Context};
+use async_trait::async_trait;
+use aws_sdk_sqs::config::{BehaviorVersion, Builder, Region, SharedAsyncSleep};
+use aws_sdk_sqs::types::{DeleteMessageBatchRequestEntry, MessageSystemAttributeName};
+use aws_sdk_sqs::{Client, Config};
+use itertools::Itertools;
+use quickwit_aws::retry::{aws_retry, AwsRetryable};
+use quickwit_aws::{get_aws_config, DEFAULT_AWS_REGION};
+use quickwit_common::rate_limited_error;
+use quickwit_common::retry::RetryParams;
+use quickwit_storage::OwnedBytes;
+use regex::Regex;
+
+use super::message::MessageMetadata;
+use super::{Queue, RawMessage};
+
+#[derive(Debug)]
+pub struct SqsQueue {
+    sqs_client: Client,
+    queue_url: String,
+    receive_retries: RetryParams,
+    acknowledge_retries: RetryParams,
+    modify_deadline_retries: RetryParams,
+}
+
+impl SqsQueue {
+    pub async fn try_new(queue_url: String) -> anyhow::Result<Self> {
+        let sqs_client = get_sqs_client(&queue_url).await?;
+        Ok(SqsQueue {
+            sqs_client,
+            queue_url,
+            receive_retries: RetryParams::standard(),
+            // Acknowledgment is retried when the message is received again
+            acknowledge_retries: RetryParams::no_retries(),
+            // Retry aggressively to avoid loosing the ownership of the message
+            modify_deadline_retries: RetryParams::aggressive(),
+        })
+    }
+}
+
+#[async_trait]
+impl Queue for SqsQueue {
+    async fn receive(
+        self: Arc<Self>,
+        max_messages: usize,
+        suggested_deadline: Duration,
+    ) -> anyhow::Result<Vec<RawMessage>> {
+        // TODO: We estimate the message deadline using the start of the
+        // ReceiveMessage request. This might be overly pessimistic: the docs
+        // state that it starts when the message is returned.
+        let initial_deadline = Instant::now() + suggested_deadline;
+        let clamped_max_messages = std::cmp::min(max_messages, 10) as i32;
+        let receive_output = aws_retry(&self.receive_retries, || async {
+            self.sqs_client
+                .receive_message()
+                .queue_url(&self.queue_url)
+                .message_system_attribute_names(MessageSystemAttributeName::ApproximateReceiveCount)
+                .wait_time_seconds(20)
+                .set_max_number_of_messages(Some(clamped_max_messages))
+                .visibility_timeout(suggested_deadline.as_secs() as i32)
+                .send()
+                .await
+        })
+        .await?;
+
+        receive_output
+            .messages
+            .unwrap_or_default()
+            .into_iter()
+            .map(|msg| {
+                let delivery_attempts: usize = msg
+                    .attributes
+                    .as_ref()
+                    .and_then(|attrs| {
```

Comment:
nitpick

Generally speaking we try to avoid leaning too much into this iterator/functional style.
When dealing with errors. It rapidly makes the code unreadable.